Add default options for Angular, add some tests from Angular

all3fox opened the pull request:
Currently this PR most surely breaks the package. It shows an attempt at more flexible default options.

The biggest change is that I made `validate.js` export more than just one function. Perhaps that is a poor choice but I felt like I needed that in tests to pass default options (for more than one style, without duplicating myself) to the validating function.

Anyway, having changed the export, I surely had to adjust for it in [lines 22-23](https://github.com/jzaefferer/commitplease/compare/master...all3fox:add-msg-styles?expand=1#diff-1dd241c4cd3fd1dd89c570cee98b79ddR23) of the tests. This is why I expect other places to break: I did not adjust anything else.

As for tests, old tests pass, new Angular tests expectedly fail -- I did not add any validation logic for it yet. Not all Angular tests have been added.

This PR is simply to show "coding skills" (or lack thereof :)) and in its current state should not be accepted.

jzaefferer commented:
Thanks, this is a good starting point to design a new API that allows for multiple styles. We can get back to this a litte later.

For now I'd like to bring this repo up to current standards, via #54, #55 and #56. Could address these together or one by one. Should make future work on this repo more enjoyable.

jzaefferer commented:
Now that we've landed the style changes in master, this needs a rebase.

jzaefferer commented:
Let's revive this PR. May need another rebase. Let's figure out how to validate the `[keyword]([component]): [subject]` pattern that Angular uses.

Currently this PR reuses `components` to specify keywords, but that's not using components quite as intended. 

How about adding another option, something like `task`, which includes those keywords from Angular. When enabled, it expects the component to be inside the parens of the task.

all3fox commented:
I took a punt at adding validation logic for Angular style. The basics are ready, there are some details that I should probably go and ask on their github. For example:

1. Their test file allows completely empy scope, yet there commit message history hints that they denote empty scope by (*)
2. Their tests define several exception cases that are not present in the style guide (`fixup!` and `WIP:` commits). Maybe the exception cases should be even wider? (with `semver` and `Merge branch into branch` commits)
3. Their `reverts: ` usage is not very consistent. They mostly use it like in the style guide, but there are like 5 commits from 2014-2013 that use `revert(scope): subject` instead of `rever: type(scope): subject`, yet there is a test case for that old usage.

jzaefferer commented:
> The basics are ready, there are some details that I should probably go and ask on their github

If its uncontroversial additions, you could create a PR against the style guide directly.

> Their test file allows completely empy scope, yet there commit message history hints that they denote empty scope by (*)

I guess that isn't mentioned in the style guide at all? Makes sense to suggest to add that to the guide and adjust the tests.

> Maybe the exception cases should be even wider?

I suppose the exceptions commitplease already supports are a useful guide here. Those should be made explicit as well.

> yet there is a test case for that old usage

Since commitplease doesn't look at old commits (unless explicitly told to, like [grunt-commitplease](https://github.com/rxaviers/grunt-commitplease) does), I suggest we drop those tests and stick with what's documented.

jzaefferer commented on lib/validate.js:
```diff
@@ -1,22 +1,38 @@
 var merge = require('mout/object/merge')
 var semver = require('semver')
 var subjectExceptions = /^(fixup|squash)!|^\[[^\]]+\]:/
+var styles = Object.freeze({ JQUERY: 0, ANGULAR: 1 })
```
Let's replace this with plain strings, so that the option becomes `style: "angular"`, which can be specified in `package.json`.

So drop this `styles` object entirely and replace the relevant check with just `style === "angular"`.

jzaefferer commented:
`lib.validate.js` is getting long enough that we could try to split the project specific validation into separate modules. May also make sense to test the smaller functions individually.

all3fox commented:
I have taken a look at angular.js issue/pull request --- they look really busy. So, I have decided not to vex them with small questions about the style guide. Instead, here is my plan:

1. I have finished implementing support for AngularJS style guide in commitplease. For that, as you have suggested, I have changed the structure of commitplease quite a bit.
2. Please review the code and comment. Sorry for dumping a few changes together in one commit on you.
3. Once I correct all the problems, we can make a new release.
4. Then I will make a PR against Angular JS where I will be deleting a bit of their old code, adding the new commitplease dependency and discussing the corner cases in their style guide. I think they will like it a lot more than just discussing hypothetical corner cases in their style guide :)
5. Depending on the discussion in that future PR, I will adjust settings for commitplease or, maybe, commitplease itself (maybe will need to release minor versions then, hopefully not at all)

So, here is how I decided to solve current corner cases that I listed earlier:

1. Disallow empty scope, i.e. disallow: `type: subject`, `type(): subject` and `type(     ): subject`. They have `type(*): subject` that means "empty scope" or "changes touch everything".
2. Support exactly the same exceptions as commitplease supports with jquery: wip commits, merge commits, semver commits and so on.
3. Disallow `reverts(scope): subject` usage and enforce `reverts: type(scope): subject` usage.

Current problems:

1. `defaults.js` holds the default settings. There is a lot of duplication there that I could not get rid of. There are two subproblems:
 1. Different styles can theoretically have identically named settings that control different things
 2. Making the `defaults` object flat means the `merge` logic must be more complicated
2. `validate.js` should be more paranoid about the `options` that it gets from the user
3. I think `commitplease` should talk more to the user (for example, tell them when the message is not checked because it is a special one or that a default style is chosen because they did not provide any settings). However, I do not know any good logging framework that would allow to silence that output if the user wants to. Do you think we need one?
4. Now that I have functions like `limits.js` and `tickets.js`, I should unit-test them too, not just `validate.js`. That would make `test.js` less intimidating. I should probably split `test.js` into several files and give it a directory of its own too. That is not done yet.

jzaefferer commented:
> I think commitplease should talk more to the user (for example, tell them when the message is not checked because it is a special one or that a default style is chosen because they did not provide any settings). However, I do not know any good logging framework that would allow to silence that output if the user wants to. Do you think we need one?

As long as its only a single line after a successful commit, I think we can get away without any options for suppressing the output.

I have to review the changes before I can respond to the other issues, will do that later.

jzaefferer commented on lib/defaults.js:
```diff
@@ -0,0 +1,60 @@
+var defaults = {
+  jquery: {
+    style: 'jquery',
+    noSemver: false,
```
I don't think we should add options that are disabled by default, and we don't have a usecase for disabling them.

jzaefferer replied:
Pretty much all the options that have the same default for both styles shouldn't be options at all.

all3fox replied:
I added for the following scenario: 

1. I do the Angular PR and tell the developers there that, for example, commitplease allows `[Tmp]: this is a temp commit`.
2. They dislike it and tell me they do not want it at all.
3. I have two options: fix the code and do a quick npm bump or fix the PR changing the setting.
4. So, I decided to make many settings. Probably bad call indeed.

Do I remove those then?

jzaefferer replied:
The example is a nasty one, since the `[tmp]` syntax is no git standard at all, unlike `!fixup`. I'd rather removing this exception completely.

 But in general it seems easy enough to fix stuff directly in commitplease (who's to stop you?), so yes, let's keep things minimal.

jzaefferer commented on lib/validate.js:
```diff
@@ -1,82 +1,103 @@
+  if (options === undefined || options.style === undefined) {
+    console.log(chalk.blue(
+      'Variables options or options.style are undefined, ' +
+      'use jquery style'
```
*`using`

jzaefferer commented on package.json:
```diff
@@ -1,6 +1,6 @@
 {
   "name": "commitplease",
-  "version": "2.3.1",
+  "version": "2.4.0",
```
Bumping version should only happen as part of `npm version` calls, which also creates a new commit and tags it. Never do that as part of a PR.

all3fox replied:
This bump is here because I rebased this branch (that was here since April) off the master. And the master got those important changes with ticketReferences. So, that is not a "manual" change, that is from master.

jzaefferer replied:
Gotcha, but then it shouldn't show up in this PR. Seems like something went wrong with the rebase.

all3fox replied:
Ah, I understand now. My mistake when I was tagging and releasing 2.4.0: I pushed the end result to my origin, which is all3fox/commitplease. Once this branch is merged, I will remove all3fox/commitplease and make this repo my origin.

I will push two commits (the `2.4.0` tag and some tests) directly to master here (my upstream), that will fix the confusion.

jzaefferer commented:
> So, I have decided not to vex them with small questions about the style guide.

Could call. To be really on the safe side, we can find a smaller project that also uses the same style guide for commits, and start there.

>  I think they will like it a lot more than just discussing hypothetical corner cases in their style guide :)

Agreed!

> 1. Disallow empty scope

👍 

> 2. validate.js should be more paranoid about the options that it gets from the user

I'm not sure that's necessary, but seems okay.

Already replied to 3. above.

> 4. Now that I have functions like limits.js and tickets.js, I should unit-test them too [...]

Makes sense, but can also wait a little longer. I suggest dealing with that when you have to touch those tests again, until then we've got more important things.

Overall looking good!

jzaefferer commented:
How are things going here? Its been 6 days since I reviewed the current PR.

all3fox commented:
I have finally come through my exams, sorry for the delay. I have taken your comments into account:

1. Dropped all extra output noise, there will be a message only if there is a style problem with commit message
2. Dropped all the settings, `defaults.js` is slimmer now, also dropped the "excpetions" setting.

I am ready to begin the PR to angular.js once this branch gets into master and on npm.

jzaefferer commented:
Nice. Since you rebased/amended, it'll take me a while to review - it helps to push fixup commits for new changes (stuff that I haven't reviewed before). Can still squash once I got caught up.

jzaefferer commented on lib/validate.js:
```diff
@@ -1,82 +1,75 @@
+
+  if (options === undefined || options.style === undefined) {
+    options = merge(defaults.jquery, options)
+  } else if (options.style === 'jquery') {
```
Could "merge" this condition into the previous one, since the result is the same:
```
if (options === undefined || options.style === undefined || options.style === 'jquery') {
 +    options = merge(defaults.jquery, options)
 +  } else if { ...
```

all3fox replied:
yes, but that feels a bit wrong: even though the content of the branch is the same, it serves different logical situations. But I will not bikeshed here and will squash those together :)

jzaefferer replied:
Can we specifiy `style: 'jquery'` as default? Then it should never be undefined and the single branch makes more sense.

jzaefferer replied:
Or just what I originally suggested, that's simple enough, not important anyway.

jzaefferer commented on lib/validate.js:
```diff
@@ -1,82 +1,75 @@
+    return []
+  }
+
+  if (lines[0].match(RegExp(options.wip))) {
```
Does angular have this exception? 

I don't even remember it existing for the jQuery style. http://contribute.jquery.org/commits-and-pull-requests/#commit-guidelines doesn't clarify anything there.

I still wonder if we need options for the exceptions, at least those that are actually part of git.

all3fox replied:
Do you mean the semver or the WIP commits? Their [styleguide](https://github.com/angular/angular.js/blob/master/CONTRIBUTING.md#commit) does not specify them. And I did not find semvers/WIPs in their commit messages on master. However, people do use "WIP" pull [requests](https://github.com/angular/angular.js/pull/12871), so it is probably a bad idea to prohibit them.

jzaefferer replied:
wip in particular, all the exceptions in general.

Since they use wip, seems fine to keep.

Still, in general, do we need options for these? I rather hardcode these exceptions and add options only when there a specific need.

A good way to see if we're adding too many options: Update the README with all the new options.

jzaefferer commented on lib/validate.js:
```diff
@@ -1,82 +1,75 @@
+  }
+
+  if (options.style === 'angular') {
+    angular(lines, options, errors)
```
`options.style` must be either 'jquery' or 'angular' at this point, right? Can we then do something like `styles[options.style](...)`?

jzaefferer commented:
I didn't review tests in detail, those seem like they haven't changed that much. Let's address the above, merge and release, then we can still fine tune later.

jzaefferer commented:
To clarify: Let's drop the `wip`, `merge`, `fixup` and `squash` options, and hardcode these checks. Then update readme and this should be good to land.

jzaefferer commented:
About documenting these: How about we land this without the README update, do an RC release on npm, then do the PR for Angular. If something comes up, we repeat that. When they're happy with the result, we document the new options and do a stable release.

all3fox commented:
@jzaefferer I am working on the README right now, it is just not very easy to formulate things clearly :) So, it takes a bit of time. I think it will be ready before the meeting today though.

jzaefferer commented:
Alright. Since you started, let's do the release candidate with the readme update.

all3fox commented:
I am finished with the `README.md` now, please take a look. Also, I did not amend-squash my commits this time (so, there are many `squash!` commits for later squashing now). Tell me if you prefer it that way.

jzaefferer commented on README.md:
```diff
@@ -2,100 +2,128 @@
 
-This [node.js](http://nodejs.org/) module validates git commit messages while you commit, according to the rules specified in the [jQuery Commit Guidelines](http://contribute.jquery.org/commits-and-pull-requests/#commit-guidelines). These are pretty generic, so this module should be useful for various projects that care about commit message formats.
+This [node.js](http://nodejs.org/) module makes sure your commit
+messages consistently follow one of these style guides:
```
Don't need to limit lines in prose, this linebreak can be dropped, see also http://contribute.jquery.org/style-guide/prose/

jzaefferer commented on README.md:
```diff
@@ -2,100 +2,128 @@
 [![Build Status](https://secure.travis-ci.org/jzaefferer/commitplease.png)](http://travis-ci.org/jzaefferer/commitplease)
 
-This [node.js](http://nodejs.org/) module validates git commit messages while you commit, according to the rules specified in the [jQuery Commit Guidelines](http://contribute.jquery.org/commits-and-pull-requests/#commit-guidelines). These are pretty generic, so this module should be useful for various projects that care about commit message formats.
+This [node.js](http://nodejs.org/) module makes sure your commit
```
"your git commit messages", add "git"

jzaefferer commented on README.md:
```diff
@@ -2,100 +2,128 @@
 ## Usage
 
-Just commit as usual. This modules installs a git commit-msg hook (unless suppressed, see below), automatically validating all commit messages as you enter them. Invalid messages will be rejected, with details on what's wrong and a copy of the input.
+Just commit as usual. This module is triggered by a git commit-msg
```
Let's drop the "just", should avoid that in general, same for "easy" or "simple"

jzaefferer commented on README.md:
```diff
@@ -2,100 +2,128 @@
-* `component`: The default `true` requires a component, set to `false` to skip the check.
-* `components`: A list of valid components. When a component is found, it's compared to the ones specified in this array.
-* `limits`: Line length limits, for subject and other lines.
+ * `limits.limit0` and `limits.others` are the hard limits for the
```
Why is `subject` now `limit0`?

jzaefferer replied:
Let's rename this to `firstLine`, to be more explicit, but not style guide specific.

jzaefferer commented on README.md:
```diff
@@ -110,10 +138,50 @@ When using commitplease together with [husky](https://github.com/typicode/husky)
+    "commitplease": "2.4.0",
+  },
+  "commitplease": {
+    "nohook": true,
```
This shouldn't be here

jzaefferer replied:
Let's make this example more generic in general.

jzaefferer commented on README.md:
```diff
@@ -2,100 +2,128 @@
+messages consistently follow one of these style guides:
+ 1. [jQuery Commit Guidelines][1]
+ 2. [AngularJS Commit Guidelines][2]
+ 3. Your own (if you really feel like it)
```
Lets drop the third item and add something like "You can also start with one of these and customize the validation rules."

jzaefferer commented:
I missed your last commit until now. Looks good, let's merge and releases this!
